rs/sns/root/src/lib.rs

```diff
@@ -704,12 +704,74 @@ impl SnsRootCanister {
         }
     }
 
+    pub async fn maybe_update_framework_canisters_memory_limit(
+        self_ref: &'static LocalKey<RefCell<Self>>,
+        management_canister_client: &impl ManagementCanisterClient,
+    ) {
+        if let Some(canisters) = self_ref.with(|state: &RefCell<SnsRootCanister>| {
+            let mut state = state.borrow_mut();
+            // If we have not updated the framework canisters' memory limit
+            if !state
+                .updated_framework_canisters_memory_limit
+                .unwrap_or_default()
+            {
+                state.updated_framework_canisters_memory_limit = Some(true);
+
+                // Since root doesn't own itself, we can't update root automatically
+                let governance = state.governance_canister_id();
+                let ledger = state.ledger_canister_id();
+                let swap = state.swap_canister_id();
+                let archives = state.archive_canister_ids.clone();
+                let index = state.index_canister_id();
+
+                Some(
+                    vec![governance, ledger, swap, index]
+                        .into_iter()
+                        .chain(archives.into_iter())
+                        .collect::<Vec<_>>(),
+                )
+            } else {
+                None
+            }
+        }) {
+            for canister_id in canisters {
+                // Update their settings to set wasm_memory_limit to 4GiB
+                let settings = CanisterSettings {
+                    wasm_memory_limit: Some(Nat::from(4_u64 << 30_u64)),
+                    ..Default::default()
+                };
+                if let Err(error) = management_canister_client
+                    .update_settings(UpdateSettings {
+                        canister_id,
+                        settings: settings.clone(),
+                        sender_canister_version: management_canister_client.canister_version(),
+                    })
+                    .await
+                {
+                    log!(
+                        ERROR,
+                        "Failed to update memory limit for canister {canister_id}: {error:?}"
+                    );
+                } else {
+                    log!(
+                        INFO,
+                        "Successfully changed memory limit for canister {canister_id}"
+                    );
+                }
+            }
+        }
+    }
+
     /// Runs periodic tasks that are not directly triggered by user input.
     pub async fn heartbeat(
         self_ref: &'static LocalKey<RefCell<Self>>,
+        management_canister_client: &impl ManagementCanisterClient,
         ledger_client: &impl LedgerCanisterClient,
         current_timestamp_seconds: u64,
     ) {
+        Self::maybe_update_framework_canisters_memory_limit(self_ref, management_canister_client)
```

Comment:
Would it be slightly better to have the new heartbeat code at the end (if it panics despite all the testing, we'd still execute the existing heartbeat code)?

Reply:
Closing this as it seems this code has already been tested (and it's temporary).